source/common/nighthawk_sink_client_impl.cc

```diff
@@ -0,0 +1,56 @@
+#include "common/nighthawk_sink_client_impl.h"
+
+#include "external/envoy/source/common/common/assert.h"
+
+namespace Nighthawk {
+
+absl::StatusOr<nighthawk::client::StoreExecutionResponse>
+NighthawkSinkClientImpl::StoreExecutionResponseStream(
+    nighthawk::client::NighthawkSink::StubInterface* nighthawk_sink_stub,
+    const nighthawk::client::StoreExecutionRequest& store_execution_request) const {
+  ::grpc::ClientContext context;
+  ::nighthawk::client::StoreExecutionResponse store_execution_response;
+  std::shared_ptr<::grpc::ClientWriterInterface<::nighthawk::client::StoreExecutionRequest>> stream(
+      nighthawk_sink_stub->StoreExecutionResponseStream(&context, &store_execution_response));
+  if (!stream->Write(store_execution_request)) {
+    return absl::UnavailableError("Failed to write request to the Nighthawk Sink gRPC channel.");
+  } else if (!stream->WritesDone()) {
+    return absl::InternalError("WritesDone() failed on the Nighthawk Sink gRPC channel.");
+  }
+  ::grpc::Status status = stream->Finish();
+  if (!status.ok()) {
+    return absl::Status(static_cast<absl::StatusCode>(status.error_code()), status.error_message());
+  }
+  return store_execution_response;
+}
+
+absl::StatusOr<nighthawk::client::SinkResponse> NighthawkSinkClientImpl::SinkRequestStream(
+    nighthawk::client::NighthawkSink::StubInterface& nighthawk_sink_stub,
+    const nighthawk::client::SinkRequest& sink_request) const {
+  nighthawk::client::SinkResponse response;
+
+  ::grpc::ClientContext context;
+  std::shared_ptr<::grpc::ClientReaderWriterInterface<nighthawk::client::SinkRequest,
+                                                      nighthawk::client::SinkResponse>>
+      stream(nighthawk_sink_stub.SinkRequestStream(&context));
+
+  if (!stream->Write(sink_request)) {
+    return absl::UnavailableError("Failed to write request to the Nighthawk Sink gRPC channel.");
+  } else if (!stream->WritesDone()) {
+    return absl::InternalError("WritesDone() failed on the Nighthawk Sink gRPC channel.");
+  }
+
+  bool got_response = false;
+  while (stream->Read(&response)) {
+    RELEASE_ASSERT(!got_response,
```

Comment:
I'm not sure I understand how this loop is supposed to work.

Reply:
Ah yes, this isn't obvious; the way we handle this reflects other places in the code base, e.g:
https://github.com/envoyproxy/nighthawk/blob/3881261ccc76e382109fdf94e5035b1f7f2b71e9/source/common/nighthawk_service_client_impl.cc#L28

Some more context on that:

At the proto api level we support returning a stream of results. The sink service proto api reflects this, and accepts what NighthawkService. ExecutionStream returns as a parameter (though we wrap it in StoreExecutionRequest messages for extensibility purposes). So this implies a stream, and not a single message.

Having said that, today we constrain what we return to a single message in the implementations where this is relevant. That's why we assert here, to make sure that stays put until an explicit choice is made otherwise.

Why do this? The intent of `NighthawkService. ExecutionStream` was to be able to stream intermediate updates some day. So having streams in the api's keeps the door open on streaming intermediary updates, without forcing a change the proto api.

Reply:
Thanks for the explanation.

Reply:
Can we convert a portion of this thread into an inline comment?

Reply:
Done in 3734e54217e9a615c117f73b69f213902cd82f25